### PR TITLE
bac-1778: add mtls example in java

### PR DESCRIPTION
**Pull request by renatolean**
Add a new folder containing a way to make call to the lean API using Java.

There are two options available to make the call:

- apache client that loads the certificate in memory
- rest template (from spring boot) that expects a pkcs12 file to be available to connect securely

### CODE REVIEW

**Review by mostafaleantech — APPROVED** [review]

File: mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        return new SSLCustomizer();
+    }
+
+    @Override
+    public void run(String... args) {
+        log.info("Starting");
+        try {
+            getBanks();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (illialean) [inline mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java]: log.error :P

File: mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        callUsingApacheClient(sslContext, props);
+    }
+
+    private static KeyStore createKeyStoreUsingHelperLib(Path clientCrtCertificate, Path privateKeyPem, Path publicKeyPem) throws GeneralSecurityException, IOException {
+        List<Certificate> certificates = CertificateUtils.loadCertificate(clientCrtCertificate, publicKeyPem);
+        PrivateKey privateKey = PemUtils.loadPrivateKey(privateKeyPem);
+
+        KeyStore keyStore = KeyStore.getInstance("pkcs12");
+        keyStore.load(null, null);
```
Comment (illialean) [inline mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java]: this looks suspicious, is it expected?

File: mtls-examples/java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
```
Comment (illialean) [inline mtls-examples/java/pom.xml]: me.leantech.exaple?

File: mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        keyStore.setKeyEntry("client-key", privateKey, KEY_PASSWORD.toCharArray(), certificates.toArray(new Certificate[3]));
+
+        return keyStore;
+    }
+
+    private static void callUsingApacheClient(SSLContext sslContext, Properties props) throws IOException, UnrecoverableKeyException, NoSuchAlgorithmException, KeyStoreException, KeyManagementException {
+        HttpClientBuilder httpClientBuilder = HttpClientBuilder.create().setSSLContext(sslContext);
+
+        try (CloseableHttpClient httpClient = httpClientBuilder.build()) {
+            String baseUrl = props.getProperty("base.url");
```
Comment (illialean) [inline mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java]: do you mention base.url anywhere?

File: mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    public static void main(String[] args) {
+        SpringApplication.run(MtlsUsingRestTemplate.class, args);
+    }
+
+    // Instantiating and injecting restTemplate (our http client)
+    @Bean
+    RestTemplate restTemplate(RestTemplateBuilder preconfiguredBuilder) {
+        return preconfiguredBuilder
+                .additionalInterceptors(new LoggingInterceptor())
+                .rootUri("https://api.leantech.me")
```
Comment (illialean) [inline mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java]: base.url?

File: mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    // method to call the banks endpoint
+    private void getBanks() throws IOException {
+        makeGetCall("/banks/v1", String.class);
+    }
+
+    // generic get method to call the bank with the correct headers and return the correct required object class
+    private <T> void makeGetCall(String urlPath, Class<T> responseClass) throws IOException {
+        Properties props = new Properties();
+        props.load(MtlsUsingApacheClient.class.getClassLoader().getResourceAsStream("application.properties"));
```
Comment (illialean) [inline mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java]: you are using spring already, maybe inject with value?

**Review by illialean — APPROVED** [review]

File: mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java (hunk 2 above)
Comment (renatolean) [reply mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java]: yeah, it is expected, see https://www.baeldung.com/java-keystore. Will add a comment to help the reader

File: mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java (hunk 4 above)
Comment (renatolean) [reply mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java]: yes, on line below, 69

File: mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java (hunk 1 above)
Comment (renatolean) [reply mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java]: yep, let me fix that!

File: mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java (hunk 6 above)
Comment (renatolean) [reply mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java]: yeah, good shout!

**Pull request merged**